### ISSUE TITLE
Combining composite types with differing element types

### ISSUE THREAD

**Issue by jarohen** [xtdb/crux#2448]
When we have nested polymorphic values, two different types may live in the same key of a DUV - e.g. when we take the union of `[:list :i64]` and `[:list :f64]`, we get `[:list [:union #{:i64 :f64}]]`. 

(n.b. we could have opted for `[:union #{[:list :i64] [:list :f64]}]`, which is a stronger type, but this would lead to bigger type explosions)

See commented out `test-least-upper-bound-upcast`.

Raising as a separate card to unblock the first pass of xtdb/xtdb#2320.
